src/legacy/core_plugins/console/np_ready/public/lib/ace_token_provider/token_provider.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IEditSession, TokenInfo } from 'brace';
+import { TokensProvider, Token, Position } from '../../interfaces';
+
+interface ActualTokenInfo extends TokenInfo {
+  type: string;
+}
+
+const toToken = (lineNumber: number, tokens: ActualTokenInfo[]): Token[] => {
+  let acc = '';
+  return tokens.map(token => {
+    const column = acc.length + 1;
+    acc += token.value || '';
+    return {
+      type: token.type,
+      value: token.value,
+      position: {
+        lineNumber,
+        column,
+      },
+    };
+  });
+};
+
+export class AceTokensProvider implements TokensProvider {
+  constructor(private readonly session: IEditSession) {}
+
+  getTokens(lineNumber: number): Token[] | null {
+    if (lineNumber < 1) return null;
```

Comment:
Line numbers are not zero-indexed?

EDIT: I think I see. In Ace they're zero-indexed, but in our interface they are one-indexed. Are there benefits to one-indexed?

Reply:
Right, so this logic captures an early return case where we know we are outside of the document.

`offset` vs `lineNumber` is a distinction monaco makes (e.g., https://microsoft.github.io/monaco-editor/api/interfaces/monaco.irange.html). I thought representing tokens as objects with `{... position: { lineNumber: number, column: number } }` values felt semantically like it should have a line number rather than an offset. For 0-index I'd probably want to change the position interface to make it clear that it's an offset - happy for other ideas here!

That being said, I really don't feel super strongly about this beyond that it's consistent 😄 . As long as consumers are equipped with representations that prevent them from doing `-1` conversion to 0-indexed or vice versa.

Reply:
I don't feel strongly about this one way or the other since the only consumers of this code will live inside Console, so making changes to it in the future should be straightforward if we change our minds.